core/hashbuffer.go

```diff
@@ -0,0 +1,149 @@
+// Copyright 2019 The go-ethereum Authors
+// This file is part of the go-ethereum library.
+//
+// The go-ethereum library is free software: you can redistribute it and/or modify
+// it under the terms of the GNU Lesser General Public License as published by
+// the Free Software Foundation, either version 3 of the License, or
+// (at your option) any later version.
+//
+// The go-ethereum library is distributed in the hope that it will be useful,
+// but WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
+// GNU Lesser General Public License for more details.
+//
+// You should have received a copy of the GNU Lesser General Public License
+// along with the go-ethereum library. If not, see <http://www.gnu.org/licenses/>.
+
+package core
+
+import (
+	"bytes"
+
+	"github.com/ethereum/go-ethereum/common"
+	"github.com/ethereum/go-ethereum/core/types"
+	"github.com/ethereum/go-ethereum/log"
+)
+
+// For EVM execution, we need around 256 items. We add a few more to allow reorgs.
+// For mainnet, a couple more would suffice, but a few more added for
+// testnets/private nets
+const hashBufferElems = 288
+
+// hashBuffer implements a storage for chains of hashes, intended to be used for quick lookup of blockhashes.
+// Internally, it uses an array of 8192 bytes to store 256 hashes in a circular buffer.
+// It enforces that all hashes added have a contiguous parent-child relation, and supports rollbacks
+// It is not thread-safe.
+type hashBuffer struct {
+	// The data holds 256 32-byte hashes. The hashes are sequential, but also a
+	// circular buffer.
+	// The `head` points to the position of the latest hash.
+	// The parent, if present, is located 32 bytes back.
+	// [.., .., ..., head-2 , head-1, head, oldest, ... ]
+	data [hashBufferElems]common.Hash
+
+	head uint64 // index of hash for the head block
+
+	headNumber uint64 // The blocknumber for head (the most recent block)
+	tailNumber uint64 // The blocknumber for tail (the oldest block)
+}
+
+// Create a new storage with a header in it.
+func NewHashStorage(header *types.Header) *hashBuffer {
+
+	// Since we take a header here, a hash storage can never be empty.
+	// This makes things easier later on (in Set)
+	hs := &hashBuffer{}
+	hs.head = 0
+	hs.headNumber = header.Number.Uint64()
+	hs.tailNumber = hs.headNumber
+	hash := header.Hash()
+	copy(hs.data[0][:], hash[:])
+	return hs
+
+}
+
+// Get locates the hash for the requested number, and copies the data into the
+// provided big.Int -- zeroing it in case the hash is not found
+func (hs *hashBuffer) Get(number uint64) (common.Hash, bool) {
+	var h common.Hash
```

Comment:
There's not much point in declaring this `h` variable. If the number is not present, we can return `common.Hash{}`. If it is present, we can return `hs.data[index]`. Since `common.Hash` is a byte **array**, it is always passed by value. You don't need to make copies.